Added various fixes for hybrid pressure coordinates

schlunma opened the pull request:
Added various fixes for CMIP5 and CMIP6 models concerning hybrid pressure coordinates (variables `cl`, `cli` and `clw`).

**Tasks**

-   [x] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [ ] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Related to #536, #540, #541, #542. (Note: Automatic closing of these issues is not desired; they should be closed when the issue is reported to the modelers).

schlunma commented:
Ready for review

schlunma commented:
I will add #551 to this PR and split it into smaller pieces due the already very large size of it. I will ping you once this is ready for review!

schlunma commented:
Ready for review. Still quite a lot of changes, but it's basically the same fix applied to many datasets. It's not possible to split this up even further, sorry!

zklaus commented on esmvalcore/cmor/_fixes/cmip5/bcc_csm1_1.py:
```diff
@@ -1,9 +1,65 @@
+from ..shared import fix_bounds
+
+
+class Cl(Fix):
```
Since this is a fix that is applied to very many models and is not at all specific to bcc-csm, maybe it's a good idea to put it into a file `common.py` and to apply it to bcc-csm1-1 in the same way as all the other models?

schlunma replied:
Good call, will do that!

zklaus commented on esmvalcore/cmor/_fixes/cmip5/bcc_csm1_1.py:
```diff
@@ -1,9 +1,65 @@
+            coords_to_fix.append('a')
+        except iris.exceptions.CoordinateNotFoundError:
+            coords_to_fix.append('ap')
+        fix_bounds(cl_cube, cubes, coords_to_fix)
```
I personally would move the loop out of `fix_bounds` to here, making `fix_bounds` a little more flexible. But ok as is.

zklaus commented on esmvalcore/cmor/_fixes/cmip5/bcc_csm1_1.py:
```diff
@@ -1,9 +1,65 @@
+        fix_bounds(cl_cube, cubes, coords_to_fix)
+
+        # Fix bounds for ap if only a is given in original file
+        ap_coord = cl_cube.coord(var_name='ap')
```
How does this work? Shouldn't files only have either `a` or `ap`?

schlunma replied:
Yes, the files have either a single `ap` coordinate or an `a` and `p0` coordinate. However, internally, `iris` converts the `a` and `p0` coordinate to `ap`, see

https://github.com/SciTools/iris/blob/06e9e2b5eae6e903e486ce12162c242d25ee4e5d/lib/iris/fileformats/netcdf.py#L677-L710

zklaus replied:
Ok, got you. Then let me propose something.

zklaus commented on esmvalcore/cmor/_fixes/cmip5/bcc_csm1_1_m.py:
```diff
@@ -1,12 +1,16 @@
+class Cl(BaseCl):
+    """Fixes for cl."""
+
+
```
```suggestion
from ..fix import Fix

from .bcc_csm1_1 import Cl
```
Should have the exact same effect. No need to derive a new class, I think.

Iff true (check with one model first), this applies of course to all similar files.

schlunma replied:
:+1:

zklaus commented on esmvalcore/cmor/_fixes/cmip6/bcc_csm2_mr.py:
```diff
@@ -0,0 +1,14 @@
+
+
+class Clw(Cl):
+    """Fixes for ``clw``."""
```
```suggestion
from ..cmip5.bcc_csm1_1 import Cl


Cli = Cl
Clw = Cl
```

schlunma replied:
:+1:

zklaus commented on esmvalcore/preprocessor/_derive/__init__.py:
```diff
@@ -96,9 +96,7 @@ def derive(cubes, short_name, long_name, units, standard_name=None):
-               f"this might be caused by missing input data for derivation "
-               f"('{short_name}' needs the variables "
-               f"{DerivedVariable().required}).")
+               f"this might be caused by missing input data for derivation")
```
I think the original message was potentially very helpful.
Why was this shortened? Maybe it's enough to add the `()` to turn required into a call, ala
```
DerivedVariable.required()
```
? The `()` after `DerivedVariable` should not be necessary due to it being a static method.

schlunma replied:
I also like the original one better (I was the one who wrote it :grin:), but since a few PRs `DerivedVariable.required()` needs `project` as argument, see

https://github.com/ESMValGroup/ESMValCore/blob/bb4597c338cc947990ec0c1f64e776b14a8e3370/esmvalcore/preprocessor/_derive/_baseclass.py#L10

However, since the `derive` function does not know about the `project`, we cannot use it anymore.

zklaus commented on esmvalcore/cmor/_fixes/cmip5/bcc_csm1_1.py:
```diff
@@ -1,9 +1,65 @@
+            ap_coord.rename('vertical pressure')
+            ap_coord.var_name = 'ap'
+            cl_cube.add_aux_coord(ap_coord, cl_cube.coord_dims(a_coord))
+
```
```suggestion
        # Fix bounds
        try:
            a_coord = cl_cube.coord(var_name='a')
            # ap_coord was constructed by iris without bounds
            # rebuild ap_coord
            cl_cube.remove_coord(ap_coord)
            fix_bounds(cl_cube, cubes, ['b', 'a'])
            p0_coord = cl_cube.coord(var_name='p0')
            ap_coord = a_coord * p0_coord.points[0]
            ap_coord.units = a_coord.units * p0_coord.units
            ap_coord.rename('vertical pressure')
            ap_coord.var_name = 'ap'
            cl_cube.add_aux_coord(ap_coord, cl_cube.coord_dims(a_coord))
        except iris.exceptions.CoordinateNotFoundError:
            coords_to_fix.append('ap')
            fix_bounds(cl_cube, cubes, ['b', 'ap'])
            ap_coord = cl_cube.coord(var_name='ap')

```

schlunma replied:
Unfortunately that won't work, since `fix_bounds(cl_cube, cubes, ['b', 'ap'])` will look if an `ap_bounds` variable is in `cubes`, which is not the case if the file has `a`/`p0` coordinates. Thus, it will raise an exception.

I will add your comment `# ap_coord was constructed by iris without bounds rebuild ap_coord` to the code, that will make it clearer.

schlunma commented:
@zklaus I rebased the master and added your suggestions.

Please not that I intentionally did not use

```python
from ..common import ClFixHybridPressureCoord as Cl
```

but

```python
from ..common import ClFixHybridPressureCoord

Cl = ClFixHybridPressureCoord
```

for two reasons: First, to mute Codacy and second, to make the it clearer that there is a fix for `Cl` in the file, otherwise I think that it might not be so obvious.
